src/pocketmine/item/WritableBook.php

```diff
@@ -0,0 +1,228 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types=1);
+
+namespace pocketmine\item;
+
+use pocketmine\nbt\tag\CompoundTag;
+use pocketmine\nbt\tag\IntTag;
+use pocketmine\nbt\tag\ListTag;
+use pocketmine\nbt\tag\StringTag;
+
+class WritableBook extends Item{
+
+	public function __construct(int $meta = 0){
+		parent::__construct(self::WRITABLE_BOOK, $meta, "Book & Quill");
+	}
+
+	/**
+	 * Returns whether the given page exists in this book.
+	 *
+	 * @param int $pageId
+	 *
+	 * @return bool
+	 */
+	public function pageExists(int $pageId) : bool{
+		return isset($this->getNamedTag()->pages->{$pageId});
+	}
+
+	/**
+	 * Returns a string containing the content of a page, or an empty string if the page doesn't exist.
+	 *
+	 * @param int $pageId
+	 *
+	 * @return string|null
+	 */
+	public function getPageText(int $pageId) : ?string{
+		if(!$this->pageExists($pageId)){
+			return null;
+		}
+		return $this->getNamedTag()->pages->{$pageId}->text->getValue();
+	}
+
+	/**
+	 * Sets the text of a page in the book. Adds the page if the page does not yet exist.
+	 *
+	 * @param int    $pageId
+	 * @param string $pageText
+	 *
+	 * @return bool indicating whether the page was created or not.
+	 */
+	public function setPageText(int $pageId, string $pageText) : bool{
+		$created = false;
+		if(!$this->pageExists($pageId)){
+			$this->addPage($pageId);
+			$created = true;
+		}
+
+		$namedTag = $this->getNamedTag();
+		$namedTag->pages->{$pageId}->text->setValue($pageText);
+		$this->setNamedTag($namedTag);
+
+		return $created;
+	}
+
+	/**
+	 * Adds a new page with the given text. (if given)
+	 *
+	 * @param int $pageId
+	 */
+	public function addPage(int $pageId) : void{
+		if($pageId < 0) {
+			throw new \InvalidArgumentException("Page number \"$pageId\" is out of range");
+		}
+		$namedTag = $this->getCorrectedNamedTag();
+
+		if(!isset($namedTag->pages) or !($namedTag->pages instanceof ListTag)){
+			$namedTag->pages = new ListTag("pages", []);
+		}
+
+		for($id = 0; $id <= $pageId; $id++) {
+			if(!$this->pageExists($id)) {
+				$namedTag->pages->{$id} = new CompoundTag("", [
+					new StringTag("text", ""),
+					new StringTag("photoname", "")
+				]);
+			}
+		}
+
```

Comment:
nitpick: some formatting issues in this function

Reply:
Fixed those just now.